src/annot/spec.ml

```diff
@@ -0,0 +1,378 @@
+  | Int32 i32 -> pf fmt "(i32 %i)" (Int32.to_int i32)
+  | Int64 i64 -> pf fmt "(i64 %i)" (Int64.to_int i64)
+  | Float32 f32 -> pf fmt "(f32 %a)" Float32.pp f32
+  | Float64 f64 -> pf fmt "(f64 %a)" Float64.pp f64
+  | Var ind -> pf fmt "%a" pp_indice ind
+  | ParamVar ind -> pf fmt "(param %a)" pp_indice ind
+  | GlobalVar ind -> pf fmt "(global %a)" pp_indice ind
+  | BinderVar ind -> pf fmt "(binder %a)" pp_indice ind
+  | UnOp (u, tm1) -> pf fmt "@[<hv 2>(%a@ %a)@]" pp_unop u pp_term tm1
+  | BinOp (b, tm1, tm2) ->
+    pf fmt "@[<hv 2>(%a@ %a@ %a)@]" pp_binop b pp_term tm1 pp_term tm2
+  | Result -> pf fmt "result"
+
+let rec pp_prop : type a. formatter -> a prop -> unit =
+ fun fmt -> function
+  | Const bool -> pf fmt "%a" pp_bool bool
+  | BinPred (b, tm1, tm2) ->
+    pf fmt "@[<hv 2>(%a@ %a@ %a)@]" pp_binpred b pp_term tm1 pp_term tm2
+  | UnConnect (u, pr1) -> pf fmt "@[<hv 2>(%a@ %a)@]" pp_unconnect u pp_prop pr1
+  | BinConnect (b, pr1, pr2) ->
+    pf fmt "@[<hv 2>(%a@ %a@ %a)@]" pp_binconnect b pp_prop pr1 pp_prop pr2
+  | Binder (b, bt, id_opt, pr1) -> (
+    match id_opt with
+    | Some id ->
+      pf fmt "@[<hv 2>(%a %a:%a@ %a)@]" pp_binder b pp_id id pp_binder_type bt
+        pp_prop pr1
+    | None ->
+      pf fmt "@[<hv 2>(%a %a@ %a)@]" pp_binder b pp_binder_type bt pp_prop pr1 )
+
+let valid_text_indice_char = function
+  | '0' .. '9'
+  | 'a' .. 'z'
+  | 'A' .. 'Z'
+  | '!' | '#' | '$' | '%' | '&' | '\'' | '*' | '+' | '-' | '.' | '/' | ':' | '<'
+  | '=' | '>' | '?' | '@' | '\\' | '^' | '_' | '`' | '|' | '~' ->
+    true
+  | _ -> false
+
+let parse_text_id id =
+  try
+    let len = String.length id in
+    let hd = String.get id 0 in
+    let tl = String.sub id 1 (len - 1) in
+    if Char.equal hd '$' && String.for_all valid_text_indice_char id then
+      Some tl
+    else None
+  with Invalid_argument _ -> None
+
+let parse_text_id_result id =
+  try
+    let len = String.length id in
+    let hd = String.get id 0 in
+    let tl = String.sub id 1 (len - 1) in
+    if Char.equal hd '$' && String.for_all valid_text_indice_char id then Ok tl
+    else Error (`Invalid_text_indice id)
+  with Invalid_argument _ -> Error (`Invalid_text_indice id)
+
+let parse_raw_id id =
+  match int_of_string id with Some id -> Some id | None -> None
+
+let parse_text_indice id =
+  match parse_text_id id with
+  | Some id -> Ok (Text id)
+  | None -> Error (`Invalid_text_indice id)
+
+let parse_indice id =
+  match (parse_text_id id, parse_raw_id id) with
+  | Some id, _ -> Ok (Text id)
+  | _, Some id -> Ok (Raw id)
+  | _, _ -> Error (`Invalid_indice id)
+
+let parse_binder_type =
+  let open Sexp in
+  function
+  | Atom "i32" -> ok I32
+  | Atom "i64" -> ok I64
+  | Atom "f32" -> ok F32
+  | Atom "f64" -> ok F64
+  | _ as bt -> Error (`Unknown_binder_type bt)
+
+let rec parse_term =
+  let open Sexp in
+  function
+  (* Int32 *)
+  | Atom i32 when Option.is_some (Int32.of_string_opt i32) ->
+    ok @@ Int32 (Int32.of_string i32)
```

Comment:
I'm not fond of doing two times the `of_string` conversion and having the two `Atom` cases in different places. What about something like:

```ocaml
| Atom a as tm -> begin
  match Int32.of_string_opt a with
  | Some i32 -> Ok (Int32 i32)
  | None ->
  match parse_tex_id a with
  | Some ind -> Ok (Var ind)
  | None ->
  if String.equal "result" a then Ok Result
  else Error (`Unknown_term tm)
end
```

?

Reply:
I think it's good idea to group `Atom` cases together. It's clearer